tests/stub/driver_parameters/test_bookmark_manager.py

```diff
@@ -474,16 +474,182 @@ def test_should_resolve_database_name_with_system_bookmarks(self):
+        tx1.commit()
+        s1.close()
+
+        s2 = self._driver.session("w")
+        tx2 = s2.begin_transaction({"return_bookmark": "bm2"})
+        list(tx2.run("RETURN 1 as n"))
+        tx2.commit()
+        s2.close()
+
+        self._server.reset()
+        begin_requests = self._server.get_requests("BEGIN")
+
+        self.assertEqual(len(begin_requests), 2)
+        self.assert_begin(
+            begin_requests[0],
+            bookmarks=system_bookmarks + neo4j_bookmarks + adb_bookmarks
+        )
+        self.assert_begin(
+            begin_requests[1],
+            bookmarks=["bm1"] + system_bookmarks + neo4j_bookmarks
+            + adb_bookmarks
+        )
+
+    def test_should_call_notify_bookmarks_when_new_bookmarks_arrive(self):
+        self._start_server(self._router, "router_with_db_name.script")
+        self._start_server(self._server, "transaction_chaining.script")
+
+        adb_bookmarks = ["adb:bm1"]
+        bookmarks_consumer_calls = []
+
+        def bookmarks_consumer(db, bookmarks):
+            bookmarks_consumer_calls.append([db, bookmarks])
+
+        self._driver = self._new_driver(
+            Neo4jBookmarkManagerConfig(
+                initial_bookmarks={
+                    "adb": adb_bookmarks
+                },
+                bookmarks_consumer=bookmarks_consumer
+            )
+        )
+
+        s1 = self._driver.session("w", database="neo4j")
+        tx1 = s1.begin_transaction({"return_bookmark": "bm1"})
+        list(tx1.run("RETURN 1 as n"))
+        tx1.commit()
+        s1.close()
+
+        s2 = self._driver.session("w")
+        tx2 = s2.begin_transaction({"order": "adb"})
+        list(tx2.run("USE adb RETURN 1 as n"))
+        tx2.commit()
+        s2.close()
+
+        self.assertEqual(2, len(bookmarks_consumer_calls))
+        self.assertEqual([
+            # first tx
+            ["neo4j", ["bm1"]],
+            # second tx
+            ["adb", ["adb:bm4"]],
+        ], bookmarks_consumer_calls)
+
+    def test_should_call_consume_for_default_db(self):
+        self._start_server(self._router, "router.script")
+        self._start_server(self._server, "transaction_chaining.script")
+
+        adb_bookmarks = ["adb:bm1"]
+        bookmarks_consumer_calls = []
+
+        def bookmarks_consumer(db, bookmarks):
+            bookmarks_consumer_calls.append([db, bookmarks])
+
+        self._driver = self._new_driver(
+            Neo4jBookmarkManagerConfig(
+                initial_bookmarks={
+                    "adb": adb_bookmarks
+                },
+                bookmarks_consumer=bookmarks_consumer
+            )
         )
 
+        s1 = self._driver.session("w")
+        tx1 = s1.begin_transaction({"return_bookmark": "bm1"})
+        list(tx1.run("RETURN 1 as n"))
+        tx1.commit()
+        s1.close()
+
+        self.assertEqual(1, len(bookmarks_consumer_calls))
+        self.assertEqual([
+            # first tx
+            ["", ["bm1"]]
```

Comment:
Why do you assume that the driver defaults to `""` if the router does not resolve the db? I assume (not tested yet) in the Python driver this will be `None`.

Reply:
Javascript defaults to "". That's why.

Reply:
Yup, tested Python. There's a difference, Python passes `None`. That means that the extension points won't nicely (only if the user code has special handling for this case.) work with 4.3 servers and earlier as they don't resolve the home db. Guess I could change the driver code...

Reply:
Adjusted

https://github.com/neo4j/neo4j-python-driver/pull/771/files#diff-3ffab728b040043b673bcf3df8558d7cd834ed3c28fcfc3d216825e53d10c397R104-R110

https://github.com/neo4j/neo4j-python-driver/pull/771/files#diff-3ffab728b040043b673bcf3df8558d7cd834ed3c28fcfc3d216825e53d10c397R137-R142

Reply:
I made it more lenient.